### DISCUSSION TITLE
New Features:

### DISCUSSION THREAD

**Discussion by nealmick** [nealmick/bb#2] [Ideas]
I have more stats to track now that we have spread, so i might make separate page just with graphs.  
What features would you like to see?

**Comment by nealmick** [comment]
Some feature Ideas:
we have gain and loss, gain being the amount of points over a correct predictions.
example:  predicted 105 - 110  actual 105 - 106 currently this would add 5 to gain. 
I want to change so that there are two more values I will call them over gain and under gain
example:  predicted 105 - 110  actual 105 - 106 in this example 4 would be added to over gain and 5 to gain.
example:  predicted 105 - 106  actual 105 - 110 in this example 1 would be added to gain and 4 to under gain. 
I want to also track when the spread is wrong.  I will call this over spread and under spread.
example:  spread +2  -2  actual 105 - 110 in this example 3 would be added to under spread
example:  spread +7  -7  actual 105 - 110 in this example 2 would be added to over spread

**Reply by nealmick** [reply]
I also want to add season games played and season games won as a team input features but the api does not have that data.
Ive though about stepping through time and adding up the values myself i do have game dates but I think it would be easier to just find another data source that does has the data i want.  I found several python package api's that might have better data too.

maybe it would be easy though if i sort games by date first, to be honest i kinda think they are already sorted by date.  but then the problem still becomes how to get new values for future games, i don't want to rely on added up values.

**Comment by nealmick** [comment]
I also think it would be really great to eventually get historical spread data setup and start predicting error in spread.

**Reply by nealmick** [reply]
this is next up!

**Reply by nealmick** [reply]
Got this done today, really improved performance, getting up to 70% accuracy now!  I will try and push updates tomorrow. Also I think moving players per team from 9 to 7 helped.

**Reply by nealmick** [reply]
Also interesting data it appears as though point spread's have gotten less accurate over time.  with a lowest accuracy of only 62% during 2021

**Reply by nealmick** [reply]
Okay pushed updates for players per team, and added point spread as a input feature.  Also the training and testing datasets are now split up in csv folder.  Getting much better results now with point spread as an input feature.  I think we are actually producing positive expected value now or at-least very close.  I am working on writing some better metrics to track percent correct with spread applied.
also live server isnt updated yet.

**Comment by nealmick** [comment]
Today i got new data working both with historical and future games from nba_api python package.  The main new data input features are games played, games won, and games lost on a daily level(per season).   I will update repo either tonight or tomorrow depending on things.  I want to add point spread as input feature getting us one step closer to spread error predictions.  This is really great though we have a new data source!

**Reply by nealmick** [reply]
this was bad idea.  nba_api python package is pretty bad.  I could'nt get it to return responses in under 30 seconds and timing out every other try, tried on server too its not my ip.   Apparently they have alot of trouble with nba.com blocking request.  many issues on there repo about it.  I tried all kinds of ways to get around but i gave up and decided to use a different api instead.  thats when i found this one: 
https://rapidapi.com/tank01/api/tank01-fantasy-stats/
super easy to use and well documented.  I had been on rapid api but never really used one there api's seems to be working well.
we are rate limited to 100 request/day soo might have to generate a few more keys on that one too but its got really good data.
im just pulling current games won and current games loss for each team.
the model is also updated to use this new data for future games, atleast quick create is, and training, with dataset etc... that all works.
I will make announcement about todays update.

**Comment by nealmick** [comment]
Okay gona be tomorow before update.  I think im going to lower playersPerTeam from 9 to 7.  

I wanted to write down this feature idea before i forget:
Make a export function to gather all the profile game data, then form in csv file so that you can look at data clearly and do csv stuff...  Im thinking all the game data, like scores, spread, and predictions...  I think the button will go on profile page, and will just redirect to static url of CSV.  Shouldnt be too difficult, need to get query set of the users games and loop through them gather the data, write csv header, and csv lines.  Im not sure what happens if you try to serve a csv in django static files.

**Reply by nealmick** [reply]
Done,  I really like this feature too.  i think it adds alot of value, lets people see all the data of the games predicted, and can do alot with csv files like graphs and equations....

**Comment by nealmick** [comment]
It would be great to have a feature that would mark players as injured.  Making the players name in edit stats clickable to mark player as injured and get next best player.  I think this would be difficult to implement.  I guess i could make a url edit/remove/<playerid>/ then basically add that player to a blacklist and re run best players.  This seems difficult. if i do make this feature it probably wont be soon.

**Comment by nealmick** [comment]
Noticed a few players on miss matched teams. or atleast will be when trades go through.  2023 NBA trade deadline buzz: Feb. 9.
ill try and update playerIdByTeamID in next few days after deadline.

Kyrie Irving is officially leaving the Brooklyn Nets to join the Dallas Mavericks.

**Reply by nealmick** [reply]
remember to update rosters tomorrow or atleast before games start again

**Comment by nealmick** [comment]
Research xgboost maybe could be useful

**Comment by nealmick** [comment]
The api for todays games is not updating properly,  the api endpoints probably changed.  Hopefully it will be fixed soon.  If not i will submit issue on the repo.   They have been good about getting things updated in the past.  Here is example api url without updated scores.

https://www.balldontlie.io/api/v1/games?dates[]=2023-02-12

**Reply by nealmick** [reply]
Looks like balldontlie api is still broken.  I submitted issue but no one has responded, maybe someone can comment on my issue and be like same, and then they see it.  Anyways Im gona try to build out the export games function tomorrow, and add more metrics like expected value, and spread accuracy.

**Reply by nealmick** [reply]
Think i found the culprit of no updating scores here:

https://github.com/ynnadkrap/balldontlie/commit/fb27742d6bd056cd84b11303019e77a56f0ca7bf


committed yesterdays right when it stoped working....

**Comment by nealmick** [comment]
Messing around training the model.  Im finding that if you overfit the model really bad (over train, way too many epochs).  Then it produces better expected value.  Ive got up to 55%.  This isnt really a good solution though as the model isnt working as it should and produces extreme results too often like teams winning by 100+ points.  I want to try and and get win/loss streak features soon.

**Comment by nealmick** [comment]
Talked to the owner of the games won/loss api we are using.  Cool guy, think he is gona add streaks as-well, which would be amazing.  still would have to find source for historical data streaks, or add them up  using game dates.  I know before I didnt want to add them up, but thats for future games.  If an api keeps the current count, I have no problem adding up historical data because it will only need to be done once, not every-time you need the current count.  This would be cool.  I also want to work on accounting for mid season trades, and injured players.  I think this can be done by using the players actual playtime to determine if they played....then if they played use there season averages as usual like average min.

**Comment by nealmick** [comment]
Added win/loss streak to web app.  Need to work on creating historical streaks with dataset so we can start using streaks as an input feature.  Also looking into injuries today.

**Comment by nealmick** [comment]
Any one have ideas on how the win/loss/streaks should be displayed?  if i put dashes or slashes it looks like a date, and is kinda confusing if i put : in-between then it looks like a time.  like 10/10/2 or 10:10:2 or 10-10-2 they look like date times.  Now there is just space in-between but i dont think its very clear.... any ideas?

![Screen Shot 2023-02-15 at 12 01 12 PM](https://user-images.githubusercontent.com/33704976/219100795-75158d3d-23fd-4a9a-967a-896eb21671ec.png)

**Reply by nealmick** [reply]
maybe two spaces!

**Comment by nealmick** [comment]
Okay next up is more stats on graphs. also making a search player function to add and remove players from games.  also going to at-least make little alert that detects if players are injured, gona be difficult to automatically match the right player and remove them so at first this will probably have to be done manually. also i wanted to experiment see what expected value would look like if we only count predictions with margins greater then like 1 point.

**Reply by nealmick** [reply]
also the get data form cant get future or past spreads, only if the game is today.  maybe i can improve this in future but to be honest i dont like that form and its way better just clicking on todays games, so dont want to spend alot of time on that.

**Comment by nealmick** [comment]
Just realized that both team rosters and season averages can be updated at the same time.  going to implement this soon make it easy to update saved stats and rosters.  
https://www.balldontlie.io/home.html#stats

**Comment by nealmick** [comment]
Okay i added a bunch of stuff today.  I havent pushed any the bigger changes yet but will try to update tonight (probably tomorrow).  The updates already live include keeping scroll position for games, ive been trying to fix that for a while but it always caused other problems, this fix isnt perfect it kinda flickers wrong scroll position on first load, then updates.  so i added a nice opacity transition on page load. kinda covers it i think.   also kinda fixed the new game form, but honestly im considering just doing away with it.  Unless i can save it by making it better, i was thinking it would be alot smoother to select date first, then be shown a list of games on that date, i could probably reuse the todays games function, this is good idea!  Also the training script is updated to track 4 levels of expected values, 0 point up to 3 point margins, meaning it only counts games that have a large enough deviation from spread to prediction... only counting strong predictions against the spread.....  This really creates some spectacular results, achieving upwards of 60% over the spread, thats plenty to cover fees and more!!  So then i added the new metrics to the game django model and added some overall fields like percentages to the profiles model.  Each game now has fields specifying if the game won over the spread, the exact margin amount, and then the 3 category margins of 1, 2, and 3 point margins.  all the data is there and being updating properly now, im just not exactly sure how im gona integrate into the graphs and donut.  I might delete the donut and just add like 5 bars to the other chart, not sure.  when i figure out the charts, ill push the update with new metrics.

**Reply by nealmick** [reply]
![Screen Shot 2023-02-17 at 7 58 21 PM](https://user-images.githubusercontent.com/33704976/219822907-e3f35101-be70-47b7-9e2c-07c94961e0d3.png)

**Reply by nealmick** [reply]
Got this built today.  Works really well and got rid of the old form!  I will say there is no historical spread data still which really sucks, its would be easy to implement but the spread api only lets u get historical odds if you pay them 10$ a month.... you can still go into admin and input spread yourself.  really happy how the date picker turned out tho!  the metrics should be fixed too, tracking margins and wins against spread.  might add those to the export function too.  I want to work on the player lookup function i think it will go under the player data on the edit page.

**Comment by nealmick** [comment]
got a whole metric ton done today.  We are now gathering and tracking injured players.  Used the same api as the team games won/loss to gather injury data using there team rosters.  I then built functionality to remove players from games.  I thought this was gona be one the most difficult features yet, but really only took a few hours.  its still not perfect i know players with apostrophe not working, but ill fix that later.  Overall im really happy with the progress today! also tweaked the training and data scripts a bit, should be getting better results now!  
Oh yea almost forgot, so as of now you must remove players first and then edit stats.  If you edit any stats then remove a player, your changes will be over written.  when you remove a player its rewriting all of the data as if it was a new game, just without the blacklisted players, and not keeping any changes to stat editing... 
![Screen Shot 2023-02-19 at 4 41 11 PM](https://user-images.githubusercontent.com/33704976/219976834-4e765df2-1ffa-4c79-b361-c6ab79f5d753.png)

**Reply by nealmick** [reply]
need to make it so you cant remove players after prediction is made.  also probably should setup a limit so there has to be 7 players left atleast, now it just errors out after you remove too many players.  also need to set the hover color, a red tint should look good.

**Comment by nealmick** [comment]
Today I fixed the remove player functionality and it should work properly really happy how the feature works.  Then i updated the team rosters, also there is now a script in util for next time.  Then I built a chat room.  not the best chat ever but it will work for now, and its not the focus of this project.  I just wanted people to be able to message and ask questions share results etc... then i made public profiles to go with it, so any user can be viewed at /public-profile/username... the template is pretty bare bones right now, ill probably add more, maybe i could put all the users games below there info.  would bo cool to make games so other users can see them, and send link to specific game, see the results, or if the stats are edited etc... i think its already is setup just the csvid would have to be fixed with the game authors username not the logged in user.

**Reply by nealmick** [reply]
The project has progressed a good bit over the past few weeks.  Games are starting again on the 23rd and I am excited to see how all the new features impact accuracy.  With the new metrics being tracked, using the export function, the results are clearer then ever and can be analyzed in detail.

**Comment by nealmick** [comment]
Multiple models?

**Comment by nealmick** [comment]
Today I added a bunch more data to the admin page, both for games and profiles.  I have been thinking about different ways to implement multiple models.  A training dashboard, where you could make datasets tweak model parameters, and test different configurations.  I dont really like the idea of training on the live server, it is the cheapest of servers, and couldnt handle it.  It would be cool to run locally tho.  I think its more important to just prove accuracy and value of current model first.  On another note things should look a bit better on mobile phones and different devices, still need to fix calendar keyboard and probably some other things, but it docent look terribly bugged.  Games are also viewable without an account now.  so you could send someone a url like 
https://nbadata.cloud/predict/edit/1068/
and should be able to open with no account.  thats also the best prediction tonight i think, highest margin of like 2.2 points.  also public profiles now show the last 25 games the user predicted.  those are also publicly available with no account.
https://nbadata.cloud/public-profile/neal

**Reply by nealmick** [reply]
probably not the best game with injuries accounted for
https://nbadata.cloud/predict/edit/1071/

**Comment by nealmick** [comment]
working on building out a community center, made a list of users, could easily turn into a leaderboard too.  Also usage is getting a bit high, we are already over 100 games today,  the daily limit is 300(team data api).  monthly we are at 1640 with a limit of 2500, renewing on the 8th(spread api).  looks like we are on track to be over limit soon.  i can register more keys on the spread api, we already using 5 free keys tho.  Probably gona end up buying a 1k/day key for 10$/month on the team data api. would then want to convert and use team data api for spread, they have spread data, just not as many markets and options.  But i might talk to the owner and see if he can add historical spread, then it would be worth it.

**Comment by nealmick** [comment]
Good website for injury data
https://dknation.draftkings.com/nba-injuries-fantasy

**Comment by nealmick** [comment]
This looks like best margin game tonight
https://nbadata.cloud/predict/edit/1122/

**Reply by nealmick** [reply]
Well that one worked out well! 
this one looking good tonight:
https://nbadata.cloud/predict/edit/1221/

**Comment by nealmick** [comment]
Really happy with where the project is currently.  Most of the features i envisioned when starting this project are built out in some form.  I still want to rework the data script to add up win/loss streaks, then we can use the streak api data for input.  Also want to use the players game time to account for injured players in training data. Collecting accuracy data for expected value on future games is going well.  I think i will add margin data to the export function too, and attempt to make graphing stuff easier.  I still would like to make a dedicated stats page, maybe above in the community view, showing a bunch more graphs.  I also still dont like how the donut chart is, oh I think the best thing for the donut chart is a 6 way split of margin 1 to 3 with each margin count, that sounds good!

**Comment by nealmick** [comment]
Got alot done today.  First I made the data script sort games by date.  Then i added up all the win/loss streaks so we can now use that data as input for future games.  Then I made the data script only use players that score 1 point or more.  This effectively gets rid of injured players, and accounts for mid season trades.  This does have a some what negative effect in that some games dont have 7 players.  those games are skipped. Overall really happy with model results, seems to have improved peak results only a little, but consistency is much better, more reliable similar results everytime.

**Comment by nealmick** [comment]
made a caching system to save recent api requests, and started building a training center thing.  can make datasets, and train on them all from the web app.  For now this isnt going on live server, its cheap server and cant handle it.  The training inputs like seasons and test games have no validation or limits, so its very easy to cause errors.  would need a really strong server to run live version, maybe i can figure out how to limit resource usage when training.  Not really sustainable with any amount of concurrent users.

**Reply by nealmick** [reply]
![Screen Shot 2023-03-02 at 4 02 43 PM](https://user-images.githubusercontent.com/33704976/222552158-546a13bc-f8b0-4f4e-9d19-1b44a003c883.png)

**Reply by nealmick** [reply]
trying 64 dense relu to 32.
getting bit better results that way.  I highly recommend using lots of test games like 1000, and just messing with other variables, probably leave seasons though.

**Comment by nealmick** [comment]
Alot of updates to the new training dashboard.  There are settings for node counts, activations, and optimizer. Additionally i discovered some problems with the player data i was using to remove mid season trades, and injured players.  This was causing alot of accuracy issues for past day or two from when those changes were implemented.  I reverted changes and am using the old playerIdByTeamID rosters.  Really was worried accuracy was down alot, all fixed now! and still have the streak data as-well.  Really happy how the training center is looking.  Currently it only supports a single model to be trained and saved.... i want to change this make models belong to users.... could probably just have a folder with each username containing that users model.  That would support one model per user.  Would be nice to eventually save multiple models, maybe have like 3 slots, and allow model naming..... Really like how this dashboard is probably at sum point will add graphs make the output display better.  also i fixed an issue with the spread cache that would provide improper odds for historical games within the 60 second refresh limit.  Now if its requesting odds from a different date then the current cached odds, it will refresh data bypassing 60 second limit and getting proper odds.

**Reply by nealmick** [reply]
![Screen Shot 2023-03-03 at 2 00 28 PM](https://user-images.githubusercontent.com/33704976/222804647-b1fea7ec-688e-4993-ac07-9f6b9360f4b5.png)

**Comment by nealmick** [comment]
![Screen Shot 2023-03-04 at 8 38 07 PM](https://user-images.githubusercontent.com/33704976/222936900-52dc2057-9b2b-4c82-8948-a05e8866a78f.png)

**Reply by nealmick** [reply]
Added graphs to training view

**Comment by nealmick** [comment]
Big update.  Can train on live server now, kinda.  The server will kill your request if its too much soo keep parameter settings low... that helps overfitting anyways.  Worked on how the training data is formed, model is performing much better now.  Added Tensorflow EarlyStopping feature, I had tried it before and just seemed useless but i found a setting called restore_best_weight and that seems to help a bit.  Your training can stop early now and not complete all the epochs if the validation loss docent improve within 4 epochs.... Also added L2 kernel regularizers, seems to help a little with overfitting.  I really like the swish activation lately.  There is a default model that is in checkpoints with default settings.  Everyone starts off with default model, and can be reverted back to default with the reset model button.  Each user can have 1 model saved to there username.... if you local install could probably set up multiple users tracking multiple models.... pretty cool!

**Comment by nealmick** [comment]
I want to change the data script to input previous x games as input features.  Input past 3 games with the results for each team. This should help the model capture more momentum and recent performance.  I think a que or line should work.  track teams previous games in list then take the last 3 games as input.... would be great to make the last game against opponent team input as-well. This would probably improve performance a good bit, but also difficult to implement, no prior knowledge of like yesterday on download?  Its too many api requests, would need to ship outdated past 3 place holder games, then update cache when new games are recorded.... Dont really like that solution for download, but i think this doable for perpetually running server. keep past 3 games for each team, and use as input.

**Comment by nealmick** [comment]
Upgraded the live server from 1 cpu and 1 g ram, to 2 cpu and 2 g ram.  Server is handling training much better now.  Still possible to cause the server to drop training request(502) if you really try and stress it.  I added early stopping, regularizers, and restore weights option toggles.  Ive got to work on some documentation for the training page, maybe a few pop up text things.  I changed the min/max range on the training parameters to help keep server load down... generally i notice the server is faster now, especially when restarting.  Working on multiple model slots.  Think there is gona be 3 to start, and will be able to select which model before making prediction, and on training page.  Not sure yet how the metrics are gona work, for now its gona be all combined, but eventually could build separate accuracy metrics for each model slot tracking future games.

**Comment by nealmick** [comment]
Added a few features recently.  Each account now can have up to 3 custom models.  The server is okay at training but not excellent its still significantly underpowered for the task.  I need to do more research if i can limit cpu usage and just make it take longer so the server docent drop requests.  Each model has a bunch of settings, i recommend reading about them on the tensorflow docs, and understanding over fitting.  Also the live server is hosting a tensorboard instance http://nbadata.cloud:6006/?darkMode=true#timeseries After 100 training logs it gets to max i need to manually clear the log folder for now, maybe fix this later.  also the logs are site wide, so its not just your user training logs.  I also setup more injury data in a table below the players table, that data is json dumped to the game django model too under the name complexInjury.  Also each game tracks which model was used to make prediction(0,1,2) and is displayed at the bottom along with the other diagnostic info.

**Comment by nealmick** [comment]
I want to build an auto remove injured player function, problem is the names in the injury data dont all match the names of the season averages mainly like - or ' and special characters.  Maybe i build anyways and just let some of them be missed and then put like icon or highlight injuries that dont match.  I also want to build a predict all button under the calendar games, make it like a game but it will just say predict all.  Also debating if the redirect after the predict button is clicked should go back to dashboard, or back to the game that was just clicked.

**Comment by nealmick** [comment]
API spread data is fixed! i also built a predict all button, this button docent remove any injured players, and exclusively use's model slot 0.  Really easy to go back and predict a bunch of games now, and test models.  I think im gona work on writing a document, call it theStrategy, explain how the project works, and ill try to specifically articulate things there.

**Comment by nealmick** [comment]
Really happy with how the new model is performing, getting much better results with smaller layers.  I think the swish activation was causing some overfitting as-well.  Really easy to test models now with the predict all button, i made a get all scores function, but im not gona upload that yet, mainly because it makes too many api calls too quickly.  Worked on making an actual homepage pretty happy with where it's at, displays todays games on an infinite carousel.  Probably will see errors with that in the coming days. 

Check out the new homepage! https://nbadata.cloud/home/
<img width="1217" alt="Screen Shot 2023-03-17 at 5 49 39 PM" src="https://user-images.githubusercontent.com/33704976/226059393-ff511eb8-77ca-4798-ae0c-7811319066ec.png">

**Comment by nealmick** [comment]
Home page looks really good now! still want to add more text, not sure how that will look.  I made a search players page, shows results in a table.  Each player has a detail view as-well.  Maybe i could link the detail view on edit game view for each player that would be cool.  For now the detail player view has 2 buttons, update season average stats, and update team.  Very use-full in future when players are traded.

**Reply by nealmick** [reply]
can now click on players name's in the game edit view, and redirects to player detail view.

**Reply by nealmick** [reply]
![Screen Shot 2023-03-21 at 1 16 39 PM](https://user-images.githubusercontent.com/33704976/226689799-90e00737-ca27-44c3-816e-f758d5c64315.png)

**Comment by nealmick** [comment]
added team detail view, shows current roster. can now click on team logo image either on game edit view or player detail view, it will then redirect to team detail view. also player names are clickable in the team roster table. 
![Screen Shot 2023-03-22 at 12 11 12 PM](https://user-images.githubusercontent.com/33704976/226968205-c73aa571-ee99-4bcd-971f-00aca1893c54.png)

**Reply by nealmick** [reply]
https://nbadata.cloud/predict/team/MIL
https://nbadata.cloud/predict/player-details/15

**Comment by nealmick** [comment]
Hopefully api will be back online soon.  I build out a team list view, with all the teams and can click on them to go to team detail view.  Also improved team detail view a bit, looks better now.

**Comment by nealmick** [comment]
got alot done today, added 2021 and 2022 seasons to training dataset. also made a system so that games can be marked as bets, and then there is a separate page with bets and stats on games marked bets.  Also i think the default model is kinda overfitting, been having good luck with really small models, 32 layer1 16 layer2.  will update default model and push updates tomorrow. Also I redid the team data so we calculate and add up win/loss/streak, no longer using saved dataset, just adding up values...

**Comment by nealmick** [comment]
Got alot done today, bets page is looking fantastic.  Uses time series data to display profit, can sort by team as-well.  I want to add a time select for 30,90,180 days etc... Also maybe add bet amount input.
![Screen Shot 2023-03-31 at 3 48 04 PM](https://user-images.githubusercontent.com/33704976/229215670-f977beea-7f1a-4449-bbae-537b621d140d.png)

**Comment by nealmick** [comment]
In the future, work on improving the Model Training Center by adding the following features:

Enhance data drop options to allow dropping players per team (from 7 to 0) and specific player stats (e.g., points or defensive rebounds). This will help focus on the most relevant data and reduce noise.

Implement test game and train game seasons sliders to provide better control over the number of games in test sets and the amount of historical data used for training. This will help tailor the model to capture recent trends effectively.

Additionally, include time series data graphs for model evaluation to visually assess the model's performance over time. This will make it easier to identify patterns and areas for improvement.

**Comment by nealmick** [comment]
Created game series training chart, displayed batches of 5 games.
![Screen Shot 2023-04-11 at 1 17 40 PM](https://user-images.githubusercontent.com/33704976/231239817-b20e5acd-ce40-4104-93f9-3a70d6134e27.png)

**Comment by nealmick** [comment]
Added data feature options
![Screen Shot 2023-04-17 at 3 55 21 PM](https://user-images.githubusercontent.com/33704976/232596207-3dc6a0f3-2428-47b4-b586-3002dfed1639.png)

**Comment by nealmick** [comment]
We now gather the previous game for each team.  Each previous game is input using the top 5 players by playtime's game stats along with the game score.  3x the input features and hits the api's really hard, so i removed predict all button for now.  each historical game is now also formed with the previous game.

**Reply by nealmick** [reply]
![Screen Shot 2023-04-23 at 12 56 46 PM](https://user-images.githubusercontent.com/33704976/233853547-3206ae7e-a16f-4352-ad7b-5b5ce7ebd060.png)

**Comment by aaaclones** [comment]
👍🏼

**Comment by dexinthecity** [comment]
just became a subscriber, is there a more detailed guide explaining what each feature is for training models?

**Reply by nealmick** [reply]
Hey thanks for subscribing! 
There is a general faq page here, but not really specific to training models:
https://nbadata.cloud/predict/faq/

Some of the settings have little popup text boxes that describe what the settings do.  All of the features supported for our models directly translate to tensorflow.    
For instances the activations functions can be found here: 
https://www.tensorflow.org/api_docs/python/tf/keras/activations
Also here are the optimizers:
https://www.tensorflow.org/api_docs/python/tf/keras/optimizers

In general this project is kinda like a wrapper for tensorflow.  When you select one of the settings in the web app, all its doing is translating that to tensorflow parameters and model architecture... I will try to write some better documentation, or might make a page specifically for info on training models.

**Comment by nealmick** [comment]
Have an idea for a RLHF feature.  So im thinking it would be great to continuously train models and update on more recent games.  a button probably at the bottom of each game, when clicked will train the model on the individual game.  Which model? i may have to build another model selector, or actually if its a previous game we should just make this only available for finished games, and also just make it select the model that was used to predict the game initially.  Then we load up the saved dataset make sure we have the final scores, and train the model on the individual game... could be cool feature and help bridge the gap between historical data and future data... also ive been looking at a few different more professional looking data source's, i really think ensuring data integrity and authenticity is paramount.  There are only a few first hand data recorders like espn and nba.com, but i want to work on a model that is based mainly off of game history.  Instead of inputting the last previous game for each team, i think inputting a 5-10 game history, kinda like a sliding time window, or time series might improve accuracy.

**Comment by Wolf242** [comment]
This is awesome. I commend you. How well have the predictions been? And what are your thoughts on the various models used for prediction like whats used for https://github.com/kyleskom/NBA-Machine-Learning-Sports-Betting/ . I have found the XGBoost models to be  consistent.

**Reply by nealmick** [reply]
Hey Thanks a bunch! Still early in the season, hoping to see an improvement with some of the newer features...  Kyle has a cool project, I think its one of the more popular open source projects.  A while back i stumbled upon it looking for data sources, I've been keeping up with the project since.  At one point last season i made a fork in an attempt to merge in some of the data from this project, it was called merge data.  The result of that ended up being that i was able integrate a bunch of the data from his project over here, but I or kyle never got around to finishing that fork and making the data available on his project.  was supposed to merge player season averages, the top 7 just like this project, but never got around to it.  Im working on a different fork now, hopefully will be plug in play for kyles project, and add in some extra data on-top... more to come one this when finished.....

**Reply by Wolf242** [reply]
Oh really I didn’t know that. Well I just got your installed and I’ll be testing the models. I’ll see where I can assist you wherever I can on this project. Do you have discord or something in case I have questions?

**Reply by nealmick** [reply]
No discord yet, there is a chat on the live web-server in the profile section.  I may consider adding a full forum in the future.  I have also considered possibly integrated another project I have thats like a social media website... its available at totally.social but its not really nba related and would take alot of work to directly integrate....

**Reply by Wolf242** [reply]
Ok ill check it out. Also i noticed the stat predictions still show players who are on the injury list. Does the model account for that when predicting the spread?

**Reply by nealmick** [reply]
Great stuff!  so the injured players are not automatically removed, you must make a final decision based on the injury report.  Sometimes players end up on the injury report as "out" or "day-to-day" and its not conclusive sometimes they still play.... there is no way to 100% know the line up of a game before it happens so we try to provide tools around that. the model does account for injured player in training.  Since the game already happened we know who actually played and use that as a line up of season averages. The idea is to try and mimic this behaviour by automatically getting the top 7 players for each team and the injury report, then letting the human operator decide on the final lineup. which players to remove. you can also edit stats if you think a player will performorm better then there season averages.  Also just one thing to clarify as well, all of our models out put 2 numbers, or predict 2 numbers,  the predictions are linear integer values and represent both teams actual game scores.  like 110 to 105 thats a example prediction. but they are actually float values for more precision, all of the input values are essentially mapped to those 2 integer values throughout training. In simplest form a tensorflow model is a black box that can take x input features, apply some math, and convert them to y output features.... the input is all the stats we offer the output is the game scores.   Now Game scores are directly translated to point spreads using the margin metric.  Vegas give home team -3 but model says home wins by 4 -> margin is 1..... bolth integer values can added together for totals although we do not evaluate models currently based on vegas totalls.  Also predicted game scores can be converted to spread lines which can be converted implied money line odds.... 

Here is some more info:
https://github.com/nealmick/bb#game-page-documentation

Also there are more tooltips and info texts popups on the live server which can be a bit more intuitive!
look for the ? buttons on the live server 

<img width="187" alt="Screen Shot 2023-11-10 at 3 44 25 PM" src="https://github.com/nealmick/bb/assets/33704976/b1803135-648c-4542-9ac7-5bd836faac99">


Available here:
https://nbadata.cloud/predict/edit/8148/

**Reply by Wolf242** [reply]
Awesome thanks for the insight, ill give it a read this weekend. Do you think you can also include the odds from bet365?

**Reply by nealmick** [reply]
Ill look into it!

**Reply by nealmick** [reply]
btw did look into bet365, the api we are using for odds data does not support bet365.  I could contact the api owner and see if they can add support, not sure if its worth it tho.  Most book makers tend to have very similar lines, as soon as a different casinos start posting different lines it becomes possible to arbitrage the difference and take out multiple bets with guarantee profits.  From what ive seen odds tend to be identical except for the few edge cases.  Let me know if you really want 365 support still, as of now im not planning on implementing this.   Still possible though,  I have thought about adding an arbitrage engine with much faster and wider vegas odds support but thats gona have wait for now, plus arbitrage gets you banned super fast, i think positive ev model betting has much more longevity until the inevitable back off.

**Reply by Wolf242** [reply]
Not sure for what it's worth, you are correct the odds are similar but sometimes the lines of higher sometimes especially on the over's. I only asked because bet365 if what i primarily mostly have access to in my country unless we use our local casino. But i think you also mentioned we could input our custom lines?

**Reply by nealmick** [reply]
I agree, Custom lines are not currently support on the web ui.  It is possible to manually go in find the game csv and edit to anything...  on the training dashboard there is a checkbox for spread data, when checked the model is trained on the home team point spread from fanduel.  The vegas line is used as input.  If the spread isnt checked in the data features when you go to a game will still show all the data but at the last moment it will drop anything the model was not set to input.....  I think best option would be to just use default setup and maybe a bit of line sniping waiting for the price to match what you want it to be....
Also wondering if vegas lines were editable how would you like to see that?? Are you talking about the odds table below?? similarly to how the season averages are editable?? Would something like a text field to leave a note on a game be somthing similar, or do you actually mean edit the model spread input??

I mean in a perfect world every book keeper would be supported, and probably would be easiest to set for the model when training.  Then just load back up the saved setting and use the book keeper set when training.  Thats probably how multiple vegas lines should work in future....  A couple notes if you input 10 spread lines that are the same or similar, you get sparse data and multicollinearity which basically means the model gets confused which variables are which... So it definitely should be a model setting.  I would say the biggest road block to building that is just the fact we dont have historical vegas odds for anything other then what we are using... for live games it fetches a bunch of lines from different books, but that data only goes back like a year. where as our training data goes back 10+ years,  originally i did extract transform and load the original spread dataset my self, so it would pretty much same process again, basically getting all the historical data correlate across data sets, then find a live data source and connect it all.... Like I definitely agree this would be a nice feature and i plan on expanding the model settings more, im not sure how soon this will get done.   I think this does offer like better user experience if your using the specific vegas line being added, however there are many books, and i kinda think its like increasing the complexity without increasing the singular accuracy.  I think in general it would be better to really get the system performing great on a single vegas line before adding a bunch more.... still in the stage where trying to improve model accuracy and figure different tricks and nifty techniques, as opposed to widening support for the system....  great ideas though i really appreciate the feedback!
